### ISSUE TITLE
oci_database_db_system db_version not consistent

### ISSUE THREAD

**Issue by tr-patrick** [terraform-providers/terraform-provider-oci#767]
### Terraform Version

Terraform v0.11.13
+ provider.oci v3.24.1

### Description:
When you create a database like following
```
resource "oci_database_db_system" "prd-db11" {
  availability_domain = "${lookup(data.oci_identity_availability_domains.ADs.availability_domains[0],"name")}"
  compartment_id      = "${var.compartment_ocid}"
  database_edition    = "${var.db_edition_prd-db11}"

  db_home {
    database {
      admin_password = "${var.db_admin_password_prd-db11}"
      db_name        = "${var.db_name_prd-db11}"
      character_set  = "${var.character_set_prd-db11}"
      ncharacter_set = "${var.n_character_set_prd-db11}"
      db_workload    = "${var.db_workload_prd-db11}"
      pdb_name       = "${var.pdb_name_prd-db11}"

      db_backup_config {
        auto_backup_enabled = false
      }
    }

    db_version   = "${var.db_version_prd-db11}"
    display_name = "${var.db_home_display_name_prd-db11}"
  
  }
  disk_redundancy = "${var.db_disk_redundancy_prd-db11}"
  shape           = "${var.db_system_shape_prd-db11}"
  subnet_id        = "${oci_core_subnet.L2AD1.id}"
  ssh_public_keys = ["${var.ssh_public_key}"]
  display_name    = "${var.db_system_display_name_prd-db11}"

  hostname                = "${var.hostname_prd-db11}"
  data_storage_percentage = "${var.data_storage_percentage_prd-db11}"
  data_storage_size_in_gb = "${var.data_storage_size_in_gb_prd-db11}"
  license_model           = "${var.license_model_prd-db11}"
  node_count              = "${var.node_count_prd-db11}"

}
```

You have to enter a version. You would want the following
```
"db_version"              = "18.5.0.0.190115"
```

You will get an error stating it wants version 18.5.0.0
```
"db_version"              = "18.5.0.0"
```

You do that and the resources get's created.

The service get's created with version 18.5.0.0.190115 because it doesn't have 18.5.0.0.

When you do terraform plan again it get's troubled. It sees the version 18.5.0.0.190115 in the state file but want version 18.5.0.0 so it would destroy my resource and create a new one. 

If i change the variable again to the more detailed it says nothing to do. 

So more detail. I created it with version 18.5.0.0 and when you do a plan again it wants to destroy and create a new resource. Which would fail because it is not allowed. 
```
-/+ module.prd-db-01.oci_database_db_system.instance (new resource required)
      id:                                                          "ocid1.dbsystem.oc1.eu-frankfurt-1.abtheljtrf4ghozme44b2pz54fqahgze3yarnr7rccu2dt75yk3iz7mmzrkq" => <computed> (forces new resource)
      availability_domain:                                         "zkLG:EU-FRANKFURT-1-AD-1" => "zkLG:EU-FRANKFURT-1-AD-1"
      backup_subnet_id:                                            "" => <computed>
      cluster_name:                                                "" => <computed>
      compartment_id:                                              "ocid1.compartment.oc1..aaaaaaaaucm4yabparxxpypl4sbiy2j6gkwq2fxf5wtcvz2ahf453oxn2p7q" => "ocid1.compartment.oc1..aaaaaaaaucm4yabparxxpypl4sbiy2j6gkwq2fxf5wtcvz2ahf453oxn2p7q"
      cpu_core_count:                                              "1" => <computed>
      data_storage_percentage:                                     "40" => "40"
      data_storage_size_in_gb:                                     "256" => "256"
      database_edition:                                            "STANDARD_EDITION" => "STANDARD_EDITION"
      db_home.#:                                                   "1" => "1"
      db_home.0.database.#:                                        "1" => "1"
      db_home.0.database.0.admin_password:                         <sensitive> => <sensitive> (attribute changed)
      db_home.0.database.0.backup_id:                              "" => <computed>
      db_home.0.database.0.backup_tde_password:                    <sensitive> => <computed> (attribute changed)
      db_home.0.database.0.character_set:                          "AL32UTF8" => "AL32UTF8"
      db_home.0.database.0.connection_strings.#:                   "1" => <computed>
      db_home.0.database.0.db_backup_config.#:                     "1" => "1"
      db_home.0.database.0.db_backup_config.0.auto_backup_enabled: "" => "false"
      db_home.0.database.0.db_name:                                "cdb1" => "cdb1"
      db_home.0.database.0.db_unique_name:                         "cdb1_fra195" => <computed>
      db_home.0.database.0.db_workload:                            "OLTP" => "OLTP"
      db_home.0.database.0.freeform_tags.%:                        "0" => <computed>
      db_home.0.database.0.id:                                     "ocid1.database.oc1.eu-frankfurt-1.abtheljtznbek5jm5xzrki2nb7axvydd7iowlbm3hh4bdsujnopmhpgy644q" => <computed>
      db_home.0.database.0.lifecycle_details:                      "" => <computed>
      db_home.0.database.0.ncharacter_set:                         "AL16UTF16" => "AL16UTF16"
      db_home.0.database.0.pdb_name:                               "pdb1" => "pdb1"
      db_home.0.database.0.state:                                  "AVAILABLE" => <computed>
      db_home.0.database.0.time_created:                           "2019-05-08 09:07:42.395 +0000 UTC" => <computed>
      db_home.0.db_version:                                        "18.5.0.0.190115" => "18.5.0.0" (forces new resource)
      db_home.0.display_name:                                      "prd-db-01" => "prd-db-01"
      db_home.0.id:                                                "ocid1.dbhome.oc1.eu-frankfurt-1.abtheljtwukecxupw6r4mgjzyuv6oj4fobnbmxodbaphdsbzcqyfwibztgnq" => <computed>
      db_home.0.last_patch_history_entry_id:                       "" => <computed>
      db_home.0.state:                                             "AVAILABLE" => <computed>
      db_home.0.time_created:                                      "2019-05-08 09:07:42.386 +0000 UTC" => <computed>
      defined_tags.%:                                              "1" => "1"
      defined_tags.Transfer.Omgeving:                              "Productie" => "Productie"
      disk_redundancy:                                             "NORMAL" => "NORMAL"
      display_name:                                                "prd-db-01" => "prd-db-01"
      domain:                                                      "l3prod.klant.oraclevcn.com" => <computed>
      fault_domains.#:                                             "1" => <computed>
      freeform_tags.%:                                             "1" => "1"
      freeform_tags.ServerType:                                    "Databaseserver" => "Databaseserver"
      hostname:                                                    "prd-db-01" => "prd-db-01"
      iorm_config_cache.#:                                         "0" => <computed>
      last_patch_history_entry_id:                                 "" => <computed>
      license_model:                                               "LICENSE_INCLUDED" => "LICENSE_INCLUDED"
      lifecycle_details:                                           "" => <computed>
      listener_port:                                               "1521" => <computed>
      node_count:                                                  "1" => "1"
      reco_storage_size_in_gb:                                     "256" => <computed>
      scan_dns_record_id:                                          "ocid1.vcndnsrecord.oc1.eu-frankfurt-1.abtheljtm4kqa6w7ghkilcretwxncr6csnozckkeu3r4cfjybd6x3xklvgpq" => <computed>
      scan_ip_ids.#:                                               "0" => <computed>
      shape:                                                       "VM.Standard2.1" => "VM.Standard2.1"
      sparse_diskgroup:                                            "" => <computed>
      ssh_public_keys.#:                                           "1" => "1"
      state:                                                       "AVAILABLE" => <computed>
      subnet_id:                                                   "ocid1.subnet.oc1.eu-frankfurt-1.aaaaaaaa2grhhxy44b4fuxrff3ejgfkxnq3bwllka2nctlu7avs4r7dzypha" => "ocid1.subnet.oc1.eu-frankfurt-1.aaaaaaaa2grhhxy44b4fuxrff3ejgfkxnq3bwllka2nctlu7avs4r7dzypha"
      time_created:                                                "2019-05-08 09:07:42.346 +0000 UTC" => <computed>
      time_zone:                                                   "UTC" => <computed>
      version:                                                     "18.5.0.0.190115" => <computed>
      vip_ids.#:                                                   "0" => <computed>


Plan: 1 to add, 0 to change, 1 to destroy.

------------------------------------------------------------------------
```

**Comment by ravitandonrt** [comment]
Thank you for reporting this issue @tr-patrick and providing full details.

# Current Expected Service Behavior (05/09/2019)
We have followed up with the Database service team on this behavior, and as of now, this is the expected behavior - where-in users will receive a full version back from the service based on the patch version used to deploy the database. This version can be different from the version user supplied as indicated above by you.

# Work-around / Mitigation
Having said that, there is a workaround/mitigation that you can apply to not report this diff in Terraform and avoiding re-creation of your database due to `db_version` diff.

Please add following to inside your `oci_database_db_system` resource config to ignore the diff in `db_version` and any other properties which may be causing the diff for now using the full property name as present in the Terraform state file:
```hcl
  lifecycle {
    ignore_changes  = ["db_home.0.db_version", "db_home.0.database.0.db_backup_config.0.auto_backup_enabled"]
    prevent_destroy = false
  }
```

You can use the following commands to view the data and property names current state of your resources
```
terraform state list
terraform state show module.prd-db-01.oci_database_db_system.instance
``` 

## Explanation for `lifecycle ignore_changes`
- `db_home.0.db_version` will protect against the version mismatch due to service returning a full version instead of the user supplied version
- `db_home.0.database.0.db_backup_config.0.auto_backup_enabled` will allow you to ignore changes on auto_backup_enabled for now. I will follow up more on this in the issue #768 
- `prevent_destroy` will protect you against accidental destruction of your critical databases, we highly recommend adding it to any mission critical resources

# Longer term / Next steps
We are going to work with the Database service on how to correct this behavior so that users do not have to add this block in their Terraform config. We cannot comment on an ETA for the same right now, but will update as soon as we have a path forward.

# Feedback
Let us know if this helped solved the problem at hand, or if you are still experiencing issues with respect to `db_version` after adding this block.

**Comment by kumar-dhanagopal** [comment]
@ravitandonrt - In your code example above, shouldn't `prevent_destroy` be set to `true`?

**Comment by ravitandonrt** [comment]
@kumar-dhanagopal - good catch, thank you, my mistake. Fixing it.

**Comment by tr-patrick** [comment]
@ravitandonrt 
Thank you for the quick and full answers. 

We tried your workaround which is exactly what we needed for this. My problem is solved for now with the workaround. 

How do we continue with this issue? Do I keep it open?

**Comment by ravitandonrt** [comment]
@tr-patrick Thanks for confirming. For now we will keep it open for tracking purposes. We will close the issue once we have a confirmation from the Database service that this is being fixed in the backend.

**Comment by rcohenma** [comment]
Hello, 
We've introduced a fix for this in v3.26.0: https://github.com/terraform-providers/terraform-provider-oci/releases/tag/v3.26.0
Please give it a try.
You should no longer see a diff on subsequent applies if the service returns a db_version with the same first four numbers in the version as the ones the user provided

So with the example you provided if you provide 18.5.0.0 and the service returns 18.5.0.0.190115 you should no longer see a diff on subsequent applies.

**Comment by tr-patrick** [comment]
@rcohenma

Looking good. Thank you for the changes
